Add PubliBike integration

misialq opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds PubliBike API integration allowing retrieval of information on availability of (e)bikes in Swiss cities.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- ~This PR fixes or closes issue: fixes #~
- ~This PR is related to issue:~ 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/22128

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

misialq committed:
Introduce PubliBike integration

misialq committed:
Fix linting

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+        self.coordinator = coordinator
+        self.attributes = {}
+        self.bike_type = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self.station.name} - {self.bike_type.capitalize()}s"
```
It less code to just handle that using a shorthand attribute in the child classes

```suggestion
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        self.refresh_common_attributes()
+        self.attributes.update({"All E-bikes": len(self.station.ebikes)})
+
+        batteries = self.get_battery_levels()
+        self.attributes = {
+            k: v for k, v in self.attributes.items() if not k.startswith("Bike ")
+        }
```
How much data is stored in there? Seems like this would put some serious strain on the state engine and state history/database storage?

See also: 

![image](https://user-images.githubusercontent.com/195327/159651831-09a726db-9ac1-4481-bfdb-914b614f7058.png)

<https://developers.home-assistant.io/docs/core/entity/sensor>

misialq replied:
Hmmm, it would be a couple/dozen of bikes every time... Now that I think about it, there's probably little use in seeing that information anyways, particularly if the bikes were already filtered based on the battery level - I removed all the bikes and only left the total count.

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+        self.attributes.update(
+            {
+                "Station name": self.station.name,
+                "Station ID": self.station.stationId,
```
What would one use the station ID for?

misialq replied:
Same as above, not much use for that - removed.

frenck commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+            cv.positive_int, vol.Range(1, 100)
+        ),
+        vol.Optional(STATION_ID): cv.positive_int,
+        vol.Optional(LATITUDE): cv.latitude,
+        vol.Optional(LONGITUDE): cv.longitude,
```
We now have a location selector, which isn't used yet/documented, but a reference PR can be found here: <https://github.com/home-assistant/core/pull/68431>

misialq replied:
Would you suggest to try using it here already or rather wait and follow up later when the other PR is merged?

frenck commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(BATTERY_LIMIT, default=BATTERY_LIMIT_DEFAULT): vol.All(
+            cv.positive_int, vol.Range(1, 100)
+        ),
```
Isn't this an option? As in, is this really required to enter?

misialq replied:
Doesn't need to be - I made it optional.

frenck commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+        vol.Required(BATTERY_LIMIT, default=BATTERY_LIMIT_DEFAULT): vol.All(
+            cv.positive_int, vol.Range(1, 100)
+        ),
+        vol.Optional(STATION_ID): cv.positive_int,
+        vol.Optional(LATITUDE): cv.latitude,
+        vol.Optional(LONGITUDE): cv.longitude,
```
What is a user enters both station ID and lat/long?

misialq replied:
Station ID would have a priority and location would be ignored. Should this also be made explicit in the schema here?

frenck commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+        errors = {}
+
+        if user_input is not None:
+            station_id = user_input.get(STATION_ID)
+            if station_id is not None:
```
```suggestion
            
            if (station_id := user_input.get(STATION_ID)) is not None:
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+    station = hass.data[DOMAIN][config_entry.entry_id]["station"]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]["coordinator"]
+    battery_limit = hass.data[DOMAIN][config_entry.entry_id][BATTERY_LIMIT]
+    entities = [
+        EBikeSensor(station, coordinator, battery_limit),
+        BikeSensor(station, coordinator),
+    ]
+
+    async_add_entities(entities)
```
```suggestion
    async_add_entities(
        [
            EBikeSensor(coordinator, battery_limit),
            BikeSensor(coordinator),
        ]
    )
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the PubliBike sensors."""
+    station = hass.data[DOMAIN][config_entry.entry_id]["station"]
```
```suggestion
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+class PubliBikeSensor(CoordinatorEntity, SensorEntity):
+    """Representation of a PubliBike sensor."""
+
+    def __init__(self, station, coordinator):
```
```suggestion
    def __init__(self, coordinator):
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+class EBikeSensor(PubliBikeSensor):
+    """Representation of an E-Bike Sensor."""
+
+    def __init__(self, station, coordinator, battery_limit=None):
+        """Initialize the sensor."""
+        super().__init__(station, coordinator)
```
```suggestion
    def __init__(self, coordinator, battery_limit=None):
        """Initialize the sensor."""
        super().__init__(coordinator)
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+class BikeSensor(PubliBikeSensor):
+    """Representation of a Bike Sensor."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the sensor."""
+        super().__init__(*args, **kwargs)
```
Let's be explicit in the arguments and loose args/kwargs.

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+    @property
+    def native_value(self) -> int:
+        """Return the state of the sensor."""
+        return len(self.station.bikes)
```
```suggestion
        return len(self.coordinator.station.bikes)
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+
+    def get_battery_levels(self):
+        """Update battery level for every bike."""
+        return {f"Bike {bike.name}": bike.batteryLevel for bike in self.station.ebikes}
```
```suggestion
        return {f"Bike {bike.name}": bike.batteryLevel for bike in self.coordinator.station.ebikes}
```

frenck commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+    @property
+    def native_value(self) -> int:
+        """Return the state of the sensor."""
+        if self.battery_limit:
+            return len(
+                [
+                    bike
+                    for bike in self.station.ebikes
+                    if bike.batteryLevel >= self.battery_limit
+                ]
+            )
+        return len(self.station.ebikes)
```
It is quite inefficient to do this each time this method is called. Instead, we can pre-calculate these values once each time the coordinator updates. Reducing the run/execution of this calculation to once per coordinator update.

misialq replied:
Moved to the coordinator.

frenck commented on homeassistant/components/publibike/strings.json:
```diff
@@ -0,0 +1,18 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "PubliBike",
```
This is a brand name, it should not be translated.

```suggestion
```

frenck reviewed (changes requested):
Hi there @misialq 👋 

Thanks for opening a PR! I've left some initial comments to get the review process started. Could you take a look?

Thanks! 👍 

../Frenck

misialq committed:
Apply suggestions from code review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

misialq committed:
Review suggestions

misialq commented on homeassistant/components/publibike/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    hass.data[DOMAIN][entry.entry_id] = {
+        "coordinator": coordinator,
+        "station": station,
+        BATTERY_LIMIT: entry.data[BATTERY_LIMIT],
```
Done.

misialq commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,112 @@
+        """Update attributes dictionary."""
+        self.attributes.update(
+            {
+                "Station name": self.station.name,
```
Good point - removed.

misialq commented:
Hi @frenck!

Many thanks for your comments! I tried to address everything - there were some minor things I wasn't quite sure about though (see respective comments).

iMicknl commented on homeassistant/components/publibike/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    )
+
+    await coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        "coordinator": coordinator,
+    }
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "coordinator": coordinator,
    }
```

And will you include more items in a follow up PR? Otherwise you don't need to make it a dict, with just the coordinator. Just add the coordinator directly.

misialq replied:
Nothing more would be included - refactored accordingly.

iMicknl commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+                    _LOGGER.error(
+                        "Unable to connect to the PubliBike API: %s", str(req_err)
+                    )
+                    return self.async_show_form(
```
I think you can simplify the config flow by not returning the `self.async_show_form` here. Especially since you want to catch unknown errors as well. An example here: https://github.com/home-assistant/core/blob/93a66cc357b226389967668441000498a10453bb/homeassistant/components/overkiz/config_flow.py#L61

misialq replied:
Refactored. I am not entirely convinced about the catch-all statement, particularly that pylint is complaining about it. Is this something we accept in the Home Assistant community?

iMicknl commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, coordinator):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._attr_name = f"{coordinator.station.name} - E-bikes"
```
Should this mention available as well? So like 
```suggestion
        self._attr_name = f"{coordinator.station.name} - Available E-bikes"
```

misialq replied:
I wouldn't necessarily say so, but if you think it would make it clearer I can change.

iMicknl commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, coordinator):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._attr_name = f"{coordinator.station.name} - Bikes"
```
What does this sensor do?

misialq replied:
Any station can have both, regular bikes and/or e-bikes. For this reason, I'm setting up two sensors here, one for the regular ones (this one) and another for the e-bikes (the one above).

misialq committed:
Review cleanup

misialq commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(BATTERY_LIMIT, default=BATTERY_LIMIT_DEFAULT): vol.All(
+            cv.positive_int, vol.Range(1, 100)
+        ),
```
I must admit, I _am_ a bit confused about how this should be set up properly. I want all of these options to be available during initial configuration - when I use OptionsFlow for this one option though, it would get moved to the post-creation configuration, right? Moreover, I am not entirely sure how to make it explicit that _either_ coordinates _or_ station id should be provided (as in, one of the two is required). Any hints would be appreciated - I couldn't really find a relevant example to compare this against  :cry:

misialq replied:
OK, I finally got this working - battery and location moved to options flow. Tests updated accordingly.

misialq committed:
Add options flow

misialq commented:
Final update: I introduced an options flow and moved the battery and location settings there. All tests were updated accordingly. Thanks in advance for re-reviews!

frenck commented:
Re-opening, this is awaiting our review.

misialq committed:
Update CODEOWNERS

frenck reviewed (changes requested):
I've marked this PR, as some jobs in our CI are failing. Please check the output of those jobs to see what went wrong and adjust the PR accordingly.

Please un-draft it once the CI passes and it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

misialq committed:
Small fixes: pylint, hassfest

misialq committed:
Add one more test

misialq committed:
Fixups

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    async_add_entities([EBikeSensor(coordinator), BikeSensor(coordinator)])
+
+
+class PubliBikeSensor(CoordinatorEntity[PubliBikeDataUpdateCoordinator], SensorEntity):
```
HA nowadays supports translation entity names and their attributes. Since 2023.8 also services are translatable.

New integrations are required to add support for [name/attributes translation](https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/). Can you please adopt your code to support it?

misialq replied:
Hey @edenhaus, I had a look at the documentation you linked but tbh I'm a little bit lost as to what exactly (and where) needs to be updated... Are you maybe aware of a PR which I could look at to see an example? (I looked at a few but couldn't really find anything useful...)

edenhaus replied:
First of all, new integration needs to use the [new naming format](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations), which is also required for translating names.

See https://github.com/home-assistant/core/pull/98224 as hopefully good example :)

edenhaus replied:
The new naming format is still missing and this is required for merging

edenhaus reviewed (changes requested):
Can you please rebase and fix the merge conflicts?

misialq committed:
Merge branch 'dev' into publibike

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        self.attributes.update({"All E-bikes": len(self.coordinator.station.ebikes)})
+        return self.attributes
```
```suggestion
        return {"All E-bikes": len(self.coordinator.station.ebikes)}
```

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+        return self.coordinator.available_ebikes
+
+    @property
+    def extra_state_attributes(self):
```
```suggestion
    def extra_state_attributes(self) -> Mapping[str, Any] | None:
```

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, coordinator):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self.attributes = {}
```
Attributes are not required.
To group all sesnor together we set a device info

```suggestion
        self._attr_device_info = DeviceInfo(
            name=coordinator.station.name,
            identifiers={(DOMAIN, coordinator.station.stationId)},
            entry_type=DeviceEntryType.SERVICE
        )
```

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+        super().__init__(coordinator)
+        self.attributes = {}
+
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes."""
+        return self.attributes
```
Can be removed

edenhaus commented on homeassistant/components/publibike/sensor.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, coordinator):
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._attr_name = f"{coordinator.station.name} - Bikes"
```
The name of the station is set by the device and can be omitted.

edenhaus reviewed (changes requested):
A requested change from my last review (translation of names and attributes) is still pending.

Please fix all requested changes.

misialq committed:
Review requests

misialq committed:
Any chance to unlock the PR conversation?

frenck commented:
@misialq The converstaion was unlocked. I've seen you pushed a couple of commits, which is great.

The CI is currently failing could you take a look? And don't forget to mark it ready for review (getting it out of draft) when it is ready for review again.

../Frenck

misialq committed:
Merge branch 'dev' into publibike

misialq committed:
Some config flow fixes

misialq committed:
Fix tests

misialq committed:
Add more tests

misialq commented:
Hey @frenck, I updated the PR and fixed all the CI checks. I did mark the PR as ready to review, however I was not able to address some of the comments about translation entity names and their attributes - tbh, I just did not get it to work. Is this something that is _required_ these days or can my PR live without it? Thanks!

edenhaus commented on homeassistant/components/publibike/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+                    vol.Optional(BATTERY_LIMIT, default=BATTERY_LIMIT_DEFAULT): vol.All(
+                        cv.positive_int, vol.Range(1, 100)
+                    ),
+                    vol.Optional(LATITUDE): cv.latitude,
+                    vol.Optional(LONGITUDE): cv.longitude,
```
```suggestion
```

Can be removed as the station has always priority and you set always a station during the config flow. See lines 83-92

edenhaus commented on homeassistant/components/publibike/__init__.py:
```diff
@@ -0,0 +1,109 @@
+        )
+        if not station:
+            raise ConfigEntryError("Station does not exists anymore")
+    else:
+        location = Location(
+            latitude=entry.options.get(LATITUDE, hass.config.latitude),
+            longitude=entry.options.get(LONGITUDE, hass.config.longitude),
+        )
+        station = await hass.async_add_executor_job(
+            publi_bike.findNearestStationTo, location
+        )
```
```suggestion
```

It can be removed. See my comment in the config flow

edenhaus commented on homeassistant/components/publibike/__init__.py:
```diff
@@ -0,0 +1,109 @@
+    """Set up PubliBike integration from a config_old entry."""
+    publi_bike = PubliBike()
+
+    if station_id := entry.data.get(STATION_ID):
+        all_stations = await hass.async_add_executor_job(publi_bike.getStations)
+        station = next(
+            filter(lambda station: station.stationId == station_id, all_stations), None
+        )
+        if not station:
+            raise ConfigEntryError("Station does not exists anymore")
```
```suggestion
    all_stations = await hass.async_add_executor_job(publi_bike.getStations)
    station = next(
        filter(lambda station: station.stationId == entry.data[STATION_ID], all_stations), None
    )
    if not station:
        raise ConfigEntryError("Station does not exists anymore")
```

edenhaus commented:
> Hey @frenck, I updated the PR and fixed all the CI checks. I did mark the PR as ready to review, however I was not able to address some of the comments about translation entity names and their attributes - tbh, I just did not get it to work. Is this something that is _required_ these days or can my PR live without it? Thanks!

Please explain which problems you have and where. The new naming format is required for new integrations and therefore we can't merge it without it.

misialq committed:
Apply suggestions from code review

Co-authored-by: Robert Resch <robert@resch.dev>

misialq committed:
Move coordinator to a new submodule

misialq committed:
Fix tests

misialq committed:
Naming refactor attempt

misialq commented:
Hey @edenhaus,
so I tried adding the EntityDescriptions to the `async_setup_entry` in `setupy.py`, together with the corresponding update to the `strings.json`, but I still cannot get the names to be displayed correctly - my sensor receives the correct name of the station but both entities do not have any name (shown as `None`). What am I still missing?
